### PR TITLE
add Huber datafits

### PR DESCRIPTION
**Pull request by EnLAI111**
When calling the function ```construct_grad(X, y, w, Xw, datafit, ws)```, I get the follow error.
```
This error may have been caused by the following argument(s):
- argument 4: Cannot determine Numba type of <class '__main__.Huber'>
```
So maybe there are bugs in the class ```Huber``` somewhere, but I can't figure out where.

### CODE REVIEW

**Comment by PABannier** [comment]
Hello @EnLAI111 , you need to decorate your `Huber` datafit with `@jitclass` decorator and specify the data type of your class attributes. Look at `Quadratic` for an example!

File: skglm/datafits/Huber.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+    def generalized_support(self, w):
+        """Return a mask with non-zero coefficients."""
+        return w != 0
+
+    def alpha_max(self, gradient0):
+        """Return penalization value for which 0 is solution."""
+        return np.max(np.abs(gradient0))
+
+
```
Comment (PABannier) [inline skglm/datafits/Huber.py:59]: Add 
```
spec_L1R = [
    ('delta', float64)
]
```

File: skglm/datafits/Huber.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+class Huber(BaseDatafit):
+    """Huber datafit.
+
+    The datafit reads::
+
+    1 / n_samples * sum (f_k(y_k - (X w)_k))
```
Comment (mathurinm) [inline skglm/datafits/Huber.py:65]: minor comment: f_k doesn't depend on k so you can rename it f.

File: skglm/datafits/Huber.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    def generalized_support(self, w):
+        """Return a mask with non-zero coefficients."""
+        return w != 0
+
+    def alpha_max(self, gradient0):
+        """Return penalization value for which 0 is solution."""
+        return np.max(np.abs(gradient0))
+
+
+class Huber(BaseDatafit):
```
Comment (PABannier) [inline skglm/datafits/Huber.py:60]: ` @jitclass(spec_huber)`

File: skglm/datafits/Huber.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    def initialize(self, X, y):
+        self.Xty = X.T @ y
+        n_features = X.shape[1]
+        self.lipschitz = np.zeros(n_features, dtype=X.dtype)
+        for j in range(n_features):
+            self.lipschitz[j] = (np.where(np.abs(y) < self.delta,
+                                          X[:, j] ** 2, 0)).sum() / len(y)
+
+    def initialize_sparse(
```
Comment (mathurinm) [inline skglm/datafits/Huber.py:97]: If you don't need to fit on sparse X, you can defer the implementation to later.

**Review by mathurinm — COMMENTED** [review]
Thanks a lot for the PR @EnLAI111 ! I did a first pass of comments, don't hesitate to ping me if something is not explicit enough or if you need feedback.

File: skglm/datafits/Huber.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+    def __init__(self, delta):
+        self.delta = delta
+
+    def initialize(self, X, y):
+        self.Xty = X.T @ y
+        n_features = X.shape[1]
+        self.lipschitz = np.zeros(n_features, dtype=X.dtype)
+        for j in range(n_features):
+            self.lipschitz[j] = (np.where(np.abs(y) < self.delta,
```
[the remark this answers is not in the document]
Comment (QB3) [reply skglm/datafits/Huber.py]: I would have said  norm(X_j) ** 2 / delta.

This makes me think on how do we want to solve this optimization problem?
If we want to use algorithms like in http://proceedings.mlr.press/v84/massias18a/massias18a.pdf and in https://arxiv.org/pdf/1902.02509.pdf, I think we will need a `custom _cd_epoch` function

[the remark this answers is not in the document]
Comment (mathurinm) [reply skglm/datafits/Huber.py]: Sure for delta ? https://github.com/scikit-learn-contrib/skglm/pull/14/files#diff-518236c0559dd6839714e8c437731d42952421ed7fed1319945a7d9bbe9f315eR20

There is no optimization over delta so far so no need for this solver, but if we need to optimize over this variable a dedicated solver will be needed

**Comment by EnLAI111** [comment]
> Thanks a lot for the PR @EnLAI111 ! I did a first pass of comments, don't hesitate to ping me if something is not explicit enough or if you need feedback.

Thanks for the comments and suggestion! I have fixed the problems. Now with WeightedL1 there is no more posted errors of ```if __name__ == '__main__'```  execution.

File: skglm/datafits/huber.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    def value(self, y, w, Xw):
+        norm_1 = np.abs(y - Xw)
+        loss = np.where(norm_1 < self.delta,
+                        0.5 * norm_1 ** 2,
+                        self.delta * norm_1 - 0.5 * self.delta ** 2)
+        return np.sum(loss) / len(Xw)
+
+    def gradient_scalar(self, X, y, w, Xw, j):
+        R = y - Xw
+        if np.abs(R[j]) < self.delta:
```
[the remark this answers is not in the document]
Comment (EnLAI111) [reply skglm/datafits/huber.py:70]: Do you mean that we should consider the case where ```y``` is multi-dimension?

[the remark this answers is not in the document]
Comment (mathurinm) [reply skglm/datafits/huber.py:70]: no y is just of shape (n_sampes,)

I'm saying the current gradient is not correct. Let h the huber funcion, let h_prime it's derivative.
The correct gradient_j is something like `- sum_i=1^n_samples X[i, j] * h_prime(y[i] - X[i, :] @ w)= - sum_i=1^n_samples X[i, j] * h_prime(R[i])`

Do you confirm ?

[the remark this answers is not in the document]
Comment (EnLAI111) [reply skglm/datafits/huber.py:70]: I got it! Thanks!

File: skglm/datafits/huber.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            return (X[:, j] @ Xw - self.Xty[j]) / len(Xw)
+        else:
+            return X[:, j] @ np.sign(-R) * self.delta / len(Xw)
+
+    def gradient_scalar_sparse(self, X_data, X_indptr, X_indices, y, Xw, j):
+        grad_j = 0.
+        for i in range(X_indptr[j], X_indptr[j + 1]):
+            R = y[X_indices[i]] - Xw[X_indices[i]]
+            if np.abs(R) < self.delta:
```
[the remark this answers is not in the document]
Comment (EnLAI111) [reply skglm/datafits/huber.py:67]: I don't think so, here ```R``` is a scalar. And maybe I no longer need store ```Xty``` in ```self```?

[the remark this answers is not in the document]
Comment (EnLAI111) [reply skglm/datafits/huber.py:67]: I will change the name and avoid use the same name for different variable. Thanks!

**Comment by PABannier** [comment]
To finalize this PR, unit tests should be added. Initially I wanted to compare it to sklearn's `HuberRegressor` but it turns out sklearn optimizes an additional parameter `sigma`. Any ideas on the unit tests we could write? @mathurinm @QB3

**Comment by tomMoral** [comment]
another solution could be to  use `scipy.optimize` to solve the problem ?

**Comment by mathurinm** [comment]
It's what HuberRegressor does already, so why rewrite this code ?

**Comment by mathurinm** [comment]
Did my pass, merge when green if LGTY @QB3 @PABannier

**Comment by mathurinm** [comment]
Thanks @EnLAI111 !
